### PR TITLE
Implement `explain`

### LINKED ISSUE #3389: Implement `Explain`
### What should be done?

Implement `Explain` method for queries just like in `pgdb`.

Integration tests are not in the scope; they should be updated in #3414, #3415, #3416.

### Where?

https://github.com/FerretDB/FerretDB/blob/main/internal/backends/postgresql/collection.go

### Definition of Done

- method implemented;
- spot refactorings done.

### PR DESCRIPTION
**Pull request by noisersup**
# Description

Closes #3389.

## Readiness checklist

- [ ] I added/updated unit tests (and they pass).
- [ ] I added/updated integration/compatibility tests (and they pass).
- [ ] I added/updated comments and checked rendering.
- [x] I made spot refactorings.
- [ ] I updated user documentation.
- [ ] I ran `task all`, and it passed.
- [x] I ensured that PR title is good enough for the changelog.
- [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.
- [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/backends/postgresql/query_iterator.go [hunk 1]
```diff
@@ -47,11 +55,16 @@ type queryIterator struct {
 //
 // Nil rows are possible and return already done iterator.
 // It still should be Close'd.
-func newQueryIterator(ctx context.Context, rows pgx.Rows) types.DocumentsIterator {
+func newQueryIterator(ctx context.Context, params *iteratorParams) types.DocumentsIterator {
```
Comment (chilagrow) [inline internal/backends/postgresql/query_iterator.go:58]: Let's not bring previous implementation (hack that I did) here. It complicates things. The implementation specific to `explain` can be not exposed here.

File: internal/backends/postgresql/helpers.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+	"golang.org/x/exp/maps"
+
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// unmarshalExplain unmarshalls the plan from EXPLAIN postgreSQL command.
+// EXPLAIN result is not sjson, so it cannot be unmarshalled by sjson.Unmarshal.
+func unmarshalExplain(b []byte) (*types.Document, error) {
```
Comment (chilagrow) [inline internal/backends/postgresql/helpers.go:30]: Can we implement this similarly to rest of the code in backend by implementing `Marshal`  `Unmarshal` interface?

File: internal/backends/postgresql/collection.go [hunk 3]
```diff
@@ -245,8 +248,53 @@ func (c *collection) DeleteAll(ctx context.Context, params *backends.DeleteAllPa
...
+
+	rows, err := p.Query(ctx, q)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	iter := newQueryIterator(ctx, &iteratorParams{
+		rows:      rows,
+		unmarshal: unmarshalExplain,
+	})
```
Comment (chilagrow) [inline internal/backends/postgresql/collection.go:286]: Let's not make change to `newQueryIterator(..)` func.

Let's do it similar to how it's done in sqlite to contain explain specific things within this scope https://github.com/FerretDB/FerretDB/blob/a29e4fb87d96025c9f7376603a313742b646df7e/internal/backends/sqlite/collection.go#L268

**Review by chilagrow — COMMENTED** [review]
Just a comment about implementation :pray:

File: internal/backends/postgresql/collection.go [hunk 4]
```diff
@@ -245,8 +248,53 @@ func (c *collection) DeleteAll(ctx context.Context, params *backends.DeleteAllPa
...
+
+	defer iter.Close()
+
+	_, queryPlan, err := iter.Next()
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (noisersup) [reply internal/backends/postgresql/collection.go:291]: Yes, although as we expect to always have 1 document treat `ErrIteratorDone` as something unexpected.

File: internal/backends/postgresql/helpers.go (hunk 2 above)
Comment (noisersup) [reply internal/backends/postgresql/helpers.go:30]: I think I don't understand. Can you show an example please?

File: internal/backends/postgresql/collection.go [hunk 5]
```diff
@@ -245,8 +248,53 @@ func (c *collection) DeleteAll(ctx context.Context, params *backends.DeleteAllPa
...
+		return nil, lazyerrors.Error(err)
+	}
```
Comment (chilagrow) [inline internal/backends/postgresql/collection.go:293]: I meant simplifying as just query that one row we need and unmarshall without using query iterator. It's not necessary to iterate on one row. :hugs: 
```suggestion
	var b []byte
	err = p.QueryRow(ctx, q).Scan(&b)
	if err != nil {
		return nil, lazyerrors.Error(err)
	}

	queryPlan, err := unmarshalExplain(b)
	if err != nil {
		return nil, lazyerrors.Error(err)
	}
```

File: internal/backends/postgresql/helpers.go (hunk 2 above)
Comment (chilagrow) [reply internal/backends/postgresql/helpers.go:30]: Sorry my explanation wasn't clear :disappointed:

There is only one row, we don't need iterator. See this diff if it makes sense :pray: 
 https://github.com/FerretDB/FerretDB/pull/3465#discussion_r1343323383

File: internal/backends/postgresql/query_iterator.go (hunk 1 above)
Comment (chilagrow) [reply internal/backends/postgresql/query_iterator.go:58]: Sorry my explanation wasn't clear :disappointed:

There is only one row, we don't need iterator. See this diff if it makes sense :pray: 
 https://github.com/FerretDB/FerretDB/pull/3465#discussion_r1343323383

File: internal/backends/postgresql/collection.go (hunk 3 above)
Comment (chilagrow) [reply internal/backends/postgresql/collection.go:286]: Just clarifying https://github.com/FerretDB/FerretDB/pull/3465#discussion_r1343323383

File: internal/backends/postgresql/query_iterator.go (hunk 1 above)
Comment (noisersup) [reply internal/backends/postgresql/query_iterator.go:58]: Makes sense, thanks!

**Review by chilagrow — APPROVED** [review]
Thanks for this :hugs:

**Comment by AlekSi** [comment]
@noisersup 
> (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Milestone (`Next`), Labels, Project and project's Sprint fields.

The milestone wasn't set

**Review by AlekSi — APPROVED** [review]

**Pull request merged**